File: lib/e2ee.ts
/**
 * End-to-end encryption for friend messages, WebCrypto only (client-side).
 *
 * Each user has an ECDH P-256 keypair: the private key lives in this
 * browser's localStorage, the public key is published to profiles.public_key.
 * A per-friendship AES-GCM key is derived with ECDH (both sides derive the
 * same key), and message bodies are stored as {v, iv, ct} envelopes — the
 * server only ever sees ciphertext.
 *
 * Honest limits (documented, not hidden): keys are per-browser (a new device
 * or cleared storage means a new keypair and old messages stay unreadable
 * there), there's no forward secrecy (one static key per pair, no ratchet),
 * and — as with any web app — whoever serves the JavaScript could ship code
 * that leaks keys. "The database only holds ciphertext" is the guarantee.
 */

export interface StoredKeyPair {
  publicJwk: JsonWebKey;
  privateJwk: JsonWebKey;
}

interface Envelope {
  v: 1;
  iv: string; // base64
  ct: string; // base64
}

const KEY_STORAGE = "beanlo-e2ee-keypair";
const ECDH = { name: "ECDH", namedCurve: "P-256" } as const;

function toB64(buf: ArrayBuffer | Uint8Array): string {
  const bytes = buf instanceof Uint8Array ? buf : new Uint8Array(buf);
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}

function fromB64(s: string): Uint8Array {
  const bin = atob(s);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return bytes;
}

/** This browser's keypair, created on first use. */
export async function getOrCreateKeyPair(): Promise<StoredKeyPair> {
  const raw = localStorage.getItem(KEY_STORAGE);
  if (raw) {
    try {
      const parsed = JSON.parse(raw) as StoredKeyPair;
      if (parsed.publicJwk && parsed.privateJwk) return parsed;
    } catch {
      // fall through and regenerate
    }
  }
  const pair = await crypto.subtle.generateKey(ECDH, true, ["deriveKey"]);
  const stored: StoredKeyPair = {
    publicJwk: await crypto.subtle.exportKey("jwk", pair.publicKey),
    privateJwk: await crypto.subtle.exportKey("jwk", pair.privateKey),
  };
  localStorage.setItem(KEY_STORAGE, JSON.stringify(stored));
  return stored;
}

/** The shared AES key for one friendship — same result on both sides. */
export async function deriveSharedKey(
  myPrivateJwk: JsonWebKey,
  theirPublicJwk: JsonWebKey
): Promise<CryptoKey> {
  const priv = await crypto.subtle.importKey("jwk", myPrivateJwk, ECDH, false, [
    "deriveKey",
  ]);
  const pub = await crypto.subtle.importKey("jwk", theirPublicJwk, ECDH, false, []);
  return crypto.subtle.deriveKey(
    { name: "ECDH", public: pub },
    priv,
    { name: "AES-GCM", length: 256 },
    false,
    ["encrypt", "decrypt"]
  );
}

export async function encryptMessage(
  key: CryptoKey,
  plaintext: string
): Promise<string> {
  const iv = crypto.getRandomValues(new Uint8Array(12));
  const ct = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv },
    key,
    new TextEncoder().encode(plaintext)
  );
  const env: Envelope = { v: 1, iv: toB64(iv), ct: toB64(ct) };
  return JSON.stringify(env);
}

/** Null when the body can't be decrypted (other device's keys, tampering). */
export async function decryptMessage(
  key: CryptoKey,
  body: string
): Promise<string | null> {
  try {
    const env = JSON.parse(body) as Envelope;
    if (env.v !== 1 || !env.iv || !env.ct) return null;
    const pt = await crypto.subtle.decrypt(
      { name: "AES-GCM", iv: fromB64(env.iv) as BufferSource },
      key,
      fromB64(env.ct) as BufferSource
    );
    return new TextDecoder().decode(pt);
  } catch {
    return null;
  }
}

File: ios/Beanlo/E2EE.swift
import Foundation
import CryptoKit

/// End-to-end encryption for friend messages — interoperable with the web's
/// WebCrypto implementation (lib/e2ee.ts): ECDH P-256 keypair per device,
/// public key published as a JWK, per-friendship AES-256-GCM key derived
/// directly from the ECDH shared secret (WebCrypto's deriveKey semantics —
/// the raw x-coordinate is the key), bodies stored as {v:1, iv, ct} JSON
/// envelopes where ct = ciphertext ‖ tag. Same honest limits as the web:
/// keys are per-device, no forward secrecy.
enum E2EE {
    private static let keyDefaultsKey = "e2ee-private-key"

    struct Envelope: Codable {
        let v: Int
        let iv: String
        let ct: String
    }

    /// This device's keypair, created on first use.
    static func privateKey() -> P256.KeyAgreement.PrivateKey {
        if let b64 = UserDefaults.standard.string(forKey: keyDefaultsKey),
           let data = Data(base64Encoded: b64),
           let key = try? P256.KeyAgreement.PrivateKey(rawRepresentation: data) {
            return key
        }
        let key = P256.KeyAgreement.PrivateKey()
        UserDefaults.standard.set(key.rawRepresentation.base64EncodedString(), forKey: keyDefaultsKey)
        return key
    }

    /// Public key as the JWK JSON string published to profiles.public_key.
    static func publicJWK() -> String {
        let pub = privateKey().publicKey.x963Representation // 0x04 ‖ x ‖ y
        let x = pub[1..<33]
        let y = pub[33..<65]
        let jwk: [String: Any] = [
            "kty": "EC",
            "crv": "P-256",
            "x": base64url(Data(x)),
            "y": base64url(Data(y)),
            "ext": true,
            "key_ops": [],
        ]
        let data = try! JSONSerialization.data(withJSONObject: jwk, options: [.sortedKeys])
        return String(data: data, encoding: .utf8)!
    }

    /// Shared AES key for one friendship — same bytes on both sides.
    static func sharedKey(theirPublicJWK: String) throws -> SymmetricKey {
        guard let data = theirPublicJWK.data(using: .utf8),
              let jwk = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let xB64 = jwk["x"] as? String, let yB64 = jwk["y"] as? String,
              let x = base64urlDecode(xB64), let y = base64urlDecode(yB64),
              x.count == 32, y.count == 32
        else { throw CocoaError(.coderInvalidValue) }
        var x963 = Data([0x04])
        x963.append(x)
        x963.append(y)
        let theirKey = try P256.KeyAgreement.PublicKey(x963Representation: x963)
        let secret = try privateKey().sharedSecretFromKeyAgreement(with: theirKey)
        // WebCrypto's ECDH→AES-GCM deriveKey uses the raw shared secret
        // (the x-coordinate, 32 bytes) as the key material directly.
        return secret.withUnsafeBytes { SymmetricKey(data: Data($0)) }
    }

    static func encrypt(_ plaintext: String, key: SymmetricKey) throws -> String {
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(Data(plaintext.utf8), using: key, nonce: nonce)
        let envelope = Envelope(
            v: 1,
            iv: Data(nonce).base64EncodedString(),
            ct: (sealed.ciphertext + sealed.tag).base64EncodedString()
        )
        let data = try JSONEncoder().encode(envelope)
        return String(data: data, encoding: .utf8)!
    }

    /// Nil when undecryptable (another device's keys, tampering).
    static func decrypt(_ body: String, key: SymmetricKey) -> String? {
        guard let data = body.data(using: .utf8),
              let env = try? JSONDecoder().decode(Envelope.self, from: data),
              env.v == 1,
              let iv = Data(base64Encoded: env.iv), iv.count == 12,
              let ct = Data(base64Encoded: env.ct), ct.count > 16
        else { return nil }
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: iv),
                ciphertext: ct.dropLast(16),
                tag: ct.suffix(16)
            )
            let plain = try AES.GCM.open(box, using: key)
            return String(data: plain, encoding: .utf8)
        } catch {
            return nil
        }
    }

    private static func base64url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func base64urlDecode(_ s: String) -> Data? {
        var b64 = s.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        while b64.count % 4 != 0 { b64 += "=" }
        return Data(base64Encoded: b64)
    }
}
